### PR TITLE
backup rewriting between `sst` and text format

### PR DESCRIPTION
**Pull request by BugenZhao**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
```

### CODE REVIEW

File: components/backup_text/src/bin/rewrite/rewrite.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+            let mut iter = reader.iter();
+            iter.seek(SeekKey::Start)?; // ignore start_key in file
+
+            while iter.valid()? {
+                let key = iter.key();
+                let value = match cf {
+                    CF_WRITE => {
+                        let write = WriteRef::parse(iter.value())?;
+                        Cow::Owned(write.to_bytes())
```
Comment (NingLin-P) [inline components/backup_text/src/bin/rewrite/rewrite.rs]: It seems unnecessary to me.

File: components/backup_text/src/bin/rewrite/rewrite.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                    _ => unreachable!(),
+                };
+                writer.put_line(key, &value)?;
+
+                count += 1;
+                iter.next()?;
+            }
+
+            let reader = writer.finish_read()?;
+            drop(reader);
```
Comment (NingLin-P) [inline components/backup_text/src/bin/rewrite/rewrite.rs]: We can add another function `fn finish`

File: components/backup_text/src/bin/rewrite/main.rs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    new_path: Option<PathBuf>,
+    /// Thread concurrency for rewriting
+    #[structopt(short, long, default_value = "8")]
+    threads: usize,
+}
+
+async fn worker(opt: Opt) -> Result<()> {
+    let Opt { path, new_path, .. } = opt;
+    let storage = {
+        let backend = make_local_backend(&path);
```
Comment (NingLin-P) [inline components/backup_text/src/bin/rewrite/main.rs]: We should also support cloud backend.

File: components/tidb_query_datatype/src/def/field_type.rs [hunk 4]
```diff
@@ -147,6 +147,9 @@ bitflags! {
         /// Field can't be NULL.
         const NOT_NULL = 1;
 
+        /// Field is part of a primary key.
+        const PRI_KEY = 1 << 1;
```
Comment (NingLin-P) [inline components/tidb_query_datatype/src/def/field_type.rs:151]: Why add this.

Comment (BugenZhao) [reply components/tidb_query_datatype/src/def/field_type.rs:151]: In TiDB `util/misc.go` `func ColumnsToProto(..)`:
```go
if (pkIsHandle && mysql.HasPriKeyFlag(c.Flag)) || c.ID == model.ExtraHandleID {
	col.PkHandle = true
}
```
We need to add this bit flag to check whether the column is part of a primary key. Though it's not used by current `json`-format TextWriter, it still might be required if we switch to `csv` format.

File: components/backup_text/src/bin/rewrite/main.rs (hunk 3 above)
Comment (BugenZhao) [reply components/backup_text/src/bin/rewrite/main.rs]: My thought is to add capabilities for other storage backends in next PRs. The signature of some functions may need to be refactored.

Comment (NingLin-P) [reply components/backup_text/src/bin/rewrite/main.rs]: Plz leave a TODO comment for now

File: components/backup_text/src/rwer.rs [hunk 5]
```diff
@@ -19,13 +19,15 @@ use tipb::{ColumnInfo, TableInfo};
 use txn_types::Key;
 
 pub struct TextWriter {
-    cxt: EvalContext,
+    ctx: EvalContext,
     data_type: Option<DataType>,
     file_writer: BufWriter<File>,
+    #[allow(dead_code)]
+    table_info: TableInfo,
```
Comment (NingLin-P) [inline components/backup_text/src/rwer.rs]: How about removing it.

File: components/backup_text/src/sst_to_text.rs [hunk 6]
```diff
@@ -1,104 +1,106 @@
...
+    let key = HrIndex::decode_index_key(key)?;
+    let value = HrIndex::decode_index_value(val, ctx, columns)?;
+    Ok(to_text(HrIndex { key, value }))
 }
 
-pub fn text_to_kv(ctx: &mut EvalContext, line: &str) -> (Vec<u8>, Vec<u8>) {
+pub fn text_to_kv(
+    ctx: &mut EvalContext,
+    line: &str,
+    _columns: &HashMap<i64, ColumnInfo>,
```
Comment (NingLin-P) [inline components/backup_text/src/sst_to_text.rs]: ditto

File: components/backup_text/src/sst_to_text.rs [hunk 7]
```diff
@@ -1,104 +1,106 @@
...
+    let key = key.into_encoded();
+    let value = value.into_bytes(ctx).unwrap();
+    (key, value)
 }
 
-pub fn index_text_to_kv(ctx: &mut EvalContext, line: &str) -> (Vec<u8>, Vec<u8>) {
+pub fn index_text_to_kv(
+    ctx: &mut EvalContext,
+    line: &str,
+    _columns: &HashMap<i64, ColumnInfo>,
```
Comment (NingLin-P) [inline components/backup_text/src/sst_to_text.rs]: ditto

**Review by NingLin-P — COMMENTED** [review]
Rest LGTM

**Pull request merged**